<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Rcpi">
<title>Rcpi Quick Reference Card • Rcpi</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Rcpi Quick Reference Card">
<meta property="og:description" content="Rcpi">
<meta property="og:image" content="https://nanx.me/Rcpi/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Rcpi</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.33.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/Rcpi.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Rcpi-quickref.html">Rcpi Quick Reference Card</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/nanxstats/Rcpi/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Rcpi Quick Reference Card</h1>
                        <h4 data-toc-skip class="author">Nan Xiao &lt;<a href="https://nanx.me" class="external-link uri">https://nanx.me</a>&gt;</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/nanxstats/Rcpi/blob/HEAD/vignettes/Rcpi-quickref.Rmd" class="external-link"><code>vignettes/Rcpi-quickref.Rmd</code></a></small>
      <div class="d-none name"><code>Rcpi-quickref.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="retrieve-protein-sequence-data-from-online-databases">Retrieve protein sequence data from online databases<a class="anchor" aria-label="anchor" href="#retrieve-protein-sequence-data-from-online-databases"></a>
</h2>
<table class="table">
<caption>Table 1: Retrieving protein sequence data from various online
databases</caption>
<thead><tr class="header">
<th align="left">Function name</th>
<th align="left">Function description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">getProt()</td>
<td align="left">Retrieve protein sequence in FASTA format or PDB format
from various online databases</td>
</tr>
<tr class="even">
<td align="left">getFASTAFromUniProt()</td>
<td align="left">Retrieve protein sequence in FASTA format from
UniProt</td>
</tr>
<tr class="odd">
<td align="left">getFASTAFromKEGG()</td>
<td align="left">Retrieve protein sequence in FASTA format from
KEGG</td>
</tr>
<tr class="even">
<td align="left">getPDBFromRCSBPDB()</td>
<td align="left">Retrieve protein sequence in PDB Format from RCSB
PDB</td>
</tr>
<tr class="odd">
<td align="left">getSeqFromUniProt()</td>
<td align="left">Retrieve protein sequence from UniProt</td>
</tr>
<tr class="even">
<td align="left">getSeqFromKEGG()</td>
<td align="left">Retrieve protein sequence from KEGG</td>
</tr>
<tr class="odd">
<td align="left">getSeqFromRCSBPDB()</td>
<td align="left">Retrieve protein sequence from RCSB PDB</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="retrieve-drug-molecular-data-from-online-databases">Retrieve drug molecular data from online databases<a class="anchor" aria-label="anchor" href="#retrieve-drug-molecular-data-from-online-databases"></a>
</h2>
<table class="table">
<caption>Table 2: Retrieving drug molecular data from various online
databases</caption>
<thead><tr class="header">
<th align="left">Function name</th>
<th align="left">Function description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">getDrug()</td>
<td align="left">Retrieve drug molecules in MOL format and SMILES format
from various online databases</td>
</tr>
<tr class="even">
<td align="left">getMolFromDrugBank()</td>
<td align="left">Retrieve drug molecules in MOL format from
DrugBank</td>
</tr>
<tr class="odd">
<td align="left">getMolFromPubChem()</td>
<td align="left">Retrieve drug molecules in MOL format from PubChem</td>
</tr>
<tr class="even">
<td align="left">getMolFromChEMBL()</td>
<td align="left">Retrieve drug molecules in MOL format from ChEMBL</td>
</tr>
<tr class="odd">
<td align="left">getMolFromKEGG()</td>
<td align="left">Retrieve drug molecules in MOL format from the
KEGG</td>
</tr>
<tr class="even">
<td align="left">getMolFromCAS()</td>
<td align="left">Retrieve drug molecules in InChI format from CAS</td>
</tr>
<tr class="odd">
<td align="left">getSmiFromDrugBank()</td>
<td align="left">Retrieve drug molecules in SMILES format from
DrugBank</td>
</tr>
<tr class="even">
<td align="left">getSmiFromPubChem()</td>
<td align="left">Retrieve drug molecules in SMILES format from
PubChem</td>
</tr>
<tr class="odd">
<td align="left">getSmiFromChEMBL()</td>
<td align="left">Retrieve drug molecules in SMILES format from
ChEMBL</td>
</tr>
<tr class="even">
<td align="left">getSmiFromKEGG()</td>
<td align="left">Retrieve drug molecules in SMILES format from KEGG</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="calculate-commonly-used-protein-sequence-derived-descriptors">Calculate commonly used protein sequence derived descriptors<a class="anchor" aria-label="anchor" href="#calculate-commonly-used-protein-sequence-derived-descriptors"></a>
</h2>
<table class="table">
<caption>Table 3: Calculating commonly used protein sequence derived
descriptors</caption>
<thead><tr class="header">
<th align="left">Function name</th>
<th align="left">Descriptor name</th>
<th align="left">Descriptor group</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">extractProtAAC()</td>
<td align="left">Amino acid composition</td>
<td align="left">Amino acid composition</td>
</tr>
<tr class="even">
<td align="left">extractProtDC()</td>
<td align="left">Dipeptide composition</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">extractProtTC()</td>
<td align="left">Tripeptide composition</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">extractProtMoreauBroto()</td>
<td align="left">Normalized Moreau-Broto autocorrelation</td>
<td align="left">Autocorrelation</td>
</tr>
<tr class="odd">
<td align="left">extractProtMoran()</td>
<td align="left">Moran autocorrelation</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">extractProtGeary()</td>
<td align="left">Geary autocorrelation</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">extractProtCTDC()</td>
<td align="left">Composition</td>
<td align="left">CTD</td>
</tr>
<tr class="even">
<td align="left">extractProtCTDT()</td>
<td align="left">Transition</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">extractProtCTDD()</td>
<td align="left">Distribution</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">extractProtCTriad()</td>
<td align="left">Conjoint Triad</td>
<td align="left">Conjoint Triad</td>
</tr>
<tr class="odd">
<td align="left">extractProtSOCN()</td>
<td align="left">Sequence-order-coupling number</td>
<td align="left">Quasi-sequence-order</td>
</tr>
<tr class="even">
<td align="left">extractProtQSO()</td>
<td align="left">Quasi-sequence-order descriptors</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">extractProtPAAC()</td>
<td align="left">Pseudo-amino acid composition</td>
<td align="left">Pseudo-amino acid composition</td>
</tr>
<tr class="even">
<td align="left">extractProtAPAAC()</td>
<td align="left">Amphiphilic pseudo-amino acid composition</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">AAindex</td>
<td align="left">AAindex data of 544 physicochemical and biological
properties for 20 amino acids</td>
<td align="left">Dataset</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="generate-profile-based-protein-representations">Generate profile-based protein representations<a class="anchor" aria-label="anchor" href="#generate-profile-based-protein-representations"></a>
</h2>
<table class="table">
<caption>Table 4: Generating profile-based protein
representations</caption>
<thead><tr class="header">
<th align="left">Function name</th>
<th align="left">Function description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">extractProtPSSM()</td>
<td align="left">Compute PSSM (Position-Specific Scoring Matrix) for
given protein sequence or peptides</td>
</tr>
<tr class="even">
<td align="left">extractProtPSSMFeature()</td>
<td align="left">Profile-based protein representation derived by
PSSM</td>
</tr>
<tr class="odd">
<td align="left">extractProtPSSMAcc()</td>
<td align="left">Profile-based protein representation derived by PSSM
and auto cross covariance (ACC)</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="generate-scales-based-descriptors-for-proteochemometrics-modeling">Generate scales-based descriptors for proteochemometrics
modeling<a class="anchor" aria-label="anchor" href="#generate-scales-based-descriptors-for-proteochemometrics-modeling"></a>
</h2>
<table class="table">
<caption>Table 5: Generating scales-based descriptors for
proteochemometrics modeling</caption>
<thead><tr class="header">
<th align="left">Function name</th>
<th align="left">Descriptor class</th>
<th align="left">Derived by</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">extractPCMScales()</td>
<td align="left">Generalized scales-based descriptors derived by
principal components analysis (PCA)</td>
<td align="left">Principal components analysis</td>
</tr>
<tr class="even">
<td align="left">extractPCMPropScales()</td>
<td align="left">Generalized scales-based descriptors derived by amino
acid properties (AAindex)</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">extractPCMDescScales()</td>
<td align="left">Generalized scales-based descriptors derived by 2D and
3D molecular descriptors (Topological, WHIM, VHSE, etc.)</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">extractPCMFAScales()</td>
<td align="left">Generalized scales-based descriptors derived by factor
analysis</td>
<td align="left">Factor analysis</td>
</tr>
<tr class="odd">
<td align="left">extractPCMMDSScales()</td>
<td align="left">Generalized scales-based descriptors derived by
multidimensional scaling (MDS)</td>
<td align="left">Multidimensional scaling</td>
</tr>
<tr class="even">
<td align="left">extractPCMBLOSUM()</td>
<td align="left">Generalized BLOSUM and PAM matrix-derived
descriptors</td>
<td align="left">Substitution matrix</td>
</tr>
<tr class="odd">
<td align="left">acc()</td>
<td align="left">Auto cross covariance (ACC) for generating scales-based
descriptors of the same length</td>
<td align="left"></td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="molecular-descriptor-sets-of-the-20-amino-acids-for-generating-scales-based-descriptors">Molecular descriptor sets of the 20 amino acids for generating
scales-based descriptors<a class="anchor" aria-label="anchor" href="#molecular-descriptor-sets-of-the-20-amino-acids-for-generating-scales-based-descriptors"></a>
</h2>
<table class="table">
<caption>Table 6: Pre-calculated molecular descriptor sets of the 20
amino acids in for generating scales-based descriptors for
proteochemometrics modeling.</caption>
<thead><tr class="header">
<th align="left">Dataset name</th>
<th align="left">Dataset description</th>
<th align="left">Dimensionality</th>
<th align="left">Calculated by</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">OptAA3d</td>
<td align="left">Optimized 20 amino acids</td>
<td align="left">–</td>
<td align="left">MOE</td>
</tr>
<tr class="even">
<td align="left">AA2DACOR</td>
<td align="left">2D autocorrelations descriptors</td>
<td align="left">92</td>
<td align="left">Dragon</td>
</tr>
<tr class="odd">
<td align="left">AA3DMoRSE</td>
<td align="left">3D-MoRSE descriptors</td>
<td align="left">160</td>
<td align="left">Dragon</td>
</tr>
<tr class="even">
<td align="left">AAACF</td>
<td align="left">Atom-centred fragments descriptors</td>
<td align="left">6</td>
<td align="left">Dragon</td>
</tr>
<tr class="odd">
<td align="left">AABurden</td>
<td align="left">Burden Eigenvalues descriptors</td>
<td align="left">62</td>
<td align="left">Dragon</td>
</tr>
<tr class="even">
<td align="left">AAConn</td>
<td align="left">Connectivity indices descriptors</td>
<td align="left">33</td>
<td align="left">Dragon</td>
</tr>
<tr class="odd">
<td align="left">AAConst</td>
<td align="left">Constitutional descriptors</td>
<td align="left">23</td>
<td align="left">Dragon</td>
</tr>
<tr class="even">
<td align="left">AAEdgeAdj</td>
<td align="left">Edge adjacency indices descriptors</td>
<td align="left">97</td>
<td align="left">Dragon</td>
</tr>
<tr class="odd">
<td align="left">AAEigIdx</td>
<td align="left">Eigenvalue-based indices descriptors</td>
<td align="left">44</td>
<td align="left">Dragon</td>
</tr>
<tr class="even">
<td align="left">AAFGC</td>
<td align="left">Functional group counts descriptors</td>
<td align="left">5</td>
<td align="left">Dragon</td>
</tr>
<tr class="odd">
<td align="left">AAGeom</td>
<td align="left">Geometrical descriptors</td>
<td align="left">41</td>
<td align="left">Dragon</td>
</tr>
<tr class="even">
<td align="left">AAGETAWAY</td>
<td align="left">GETAWAY descriptors</td>
<td align="left">194</td>
<td align="left">Dragon</td>
</tr>
<tr class="odd">
<td align="left">AAInfo</td>
<td align="left">Information indices descriptors</td>
<td align="left">47</td>
<td align="left">Dragon</td>
</tr>
<tr class="even">
<td align="left">AAMolProp</td>
<td align="left">Molecular properties descriptors</td>
<td align="left">12</td>
<td align="left">Dragon</td>
</tr>
<tr class="odd">
<td align="left">AARandic</td>
<td align="left">Randic molecular profiles descriptors</td>
<td align="left">41</td>
<td align="left">Dragon</td>
</tr>
<tr class="even">
<td align="left">AARDF</td>
<td align="left">RDF descriptors</td>
<td align="left">82</td>
<td align="left">Dragon</td>
</tr>
<tr class="odd">
<td align="left">AATopo</td>
<td align="left">Topological descriptors</td>
<td align="left">78</td>
<td align="left">Dragon</td>
</tr>
<tr class="even">
<td align="left">AATopoChg</td>
<td align="left">Topological charge indices descriptors</td>
<td align="left">15</td>
<td align="left">Dragon</td>
</tr>
<tr class="odd">
<td align="left">AAWalk</td>
<td align="left">Walk and path counts descriptors</td>
<td align="left">40</td>
<td align="left">Dragon</td>
</tr>
<tr class="even">
<td align="left">AAWHIM</td>
<td align="left">WHIM descriptors</td>
<td align="left">99</td>
<td align="left">Dragon</td>
</tr>
<tr class="odd">
<td align="left">AACPSA</td>
<td align="left">CPSA descriptors</td>
<td align="left">41</td>
<td align="left">Accelrys Discovery Studio</td>
</tr>
<tr class="even">
<td align="left">AADescAll</td>
<td align="left">All the 2D descriptors calculated by Dragon</td>
<td align="left">1171</td>
<td align="left">Dragon</td>
</tr>
<tr class="odd">
<td align="left">AAMOE2D</td>
<td align="left">All the 2D descriptors calculated by MOE</td>
<td align="left">148</td>
<td align="left">MOE</td>
</tr>
<tr class="even">
<td align="left">AAMOE3D</td>
<td align="left">All the 3D descriptors calculated by MOE</td>
<td align="left">143</td>
<td align="left">MOE</td>
</tr>
<tr class="odd">
<td align="left">AABLOSUM45</td>
<td align="left">BLOSUM45 matrix for 20 amino acids</td>
<td align="left"><span class="math inline">\(20 \times 20\)</span></td>
<td align="left">Biostrings</td>
</tr>
<tr class="even">
<td align="left">AABLOSUM50</td>
<td align="left">BLOSUM50 matrix for 20 amino acids</td>
<td align="left"><span class="math inline">\(20 \times 20\)</span></td>
<td align="left">Biostrings</td>
</tr>
<tr class="odd">
<td align="left">AABLOSUM62</td>
<td align="left">BLOSUM62 matrix for 20 amino acids</td>
<td align="left"><span class="math inline">\(20 \times 20\)</span></td>
<td align="left">Biostrings</td>
</tr>
<tr class="even">
<td align="left">AABLOSUM80</td>
<td align="left">BLOSUM80 matrix for 20 amino acids</td>
<td align="left"><span class="math inline">\(20 \times 20\)</span></td>
<td align="left">Biostrings</td>
</tr>
<tr class="odd">
<td align="left">AABLOSUM100</td>
<td align="left">BLOSUM100 matrix for 20 amino acids</td>
<td align="left"><span class="math inline">\(20 \times 20\)</span></td>
<td align="left">Biostrings</td>
</tr>
<tr class="even">
<td align="left">AAPAM30</td>
<td align="left">PAM30 matrix for 20 amino acids</td>
<td align="left"><span class="math inline">\(20 \times 20\)</span></td>
<td align="left">Biostrings</td>
</tr>
<tr class="odd">
<td align="left">AAPAM40</td>
<td align="left">PAM40 matrix for 20 amino acids</td>
<td align="left"><span class="math inline">\(20 \times 20\)</span></td>
<td align="left">Biostrings</td>
</tr>
<tr class="even">
<td align="left">AAPAM70</td>
<td align="left">PAM70 matrix for 20 amino acids</td>
<td align="left"><span class="math inline">\(20 \times 20\)</span></td>
<td align="left">Biostrings</td>
</tr>
<tr class="odd">
<td align="left">AAPAM120</td>
<td align="left">PAM120 matrix for 20 amino acids</td>
<td align="left"><span class="math inline">\(20 \times 20\)</span></td>
<td align="left">Biostrings</td>
</tr>
<tr class="even">
<td align="left">AAPAM250</td>
<td align="left">PAM250 matrix for 20 amino acids</td>
<td align="left"><span class="math inline">\(20 \times 20\)</span></td>
<td align="left">Biostrings</td>
</tr>
</tbody>
</table>
<p><strong>Note</strong>: non-informative descriptors (e.g. descriptors
with only one value across all the 20 amino acids) in these datasets
have been filtered out.</p>
</div>
<div class="section level2">
<h2 id="molecular-descriptors">Molecular descriptors<a class="anchor" aria-label="anchor" href="#molecular-descriptors"></a>
</h2>
<table class="table">
<caption>Table 7: Molecular descriptors</caption>
<thead><tr class="header">
<th align="left">Function name</th>
<th align="left">Descriptor name</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">extractDrugAIO()</td>
<td align="left">All the molecular descriptors in the package</td>
</tr>
<tr class="even">
<td align="left">extractDrugALOGP()</td>
<td align="left">Atom additive logP and molar refractivity values
descriptor</td>
</tr>
<tr class="odd">
<td align="left">extractDrugAminoAcidCount()</td>
<td align="left">Number of amino acids</td>
</tr>
<tr class="even">
<td align="left">extractDrugApol()</td>
<td align="left">Sum of the atomic polarizabilities</td>
</tr>
<tr class="odd">
<td align="left">extractDrugAromaticAtomsCount()</td>
<td align="left">Number of aromatic atoms</td>
</tr>
<tr class="even">
<td align="left">extractDrugAromaticBondsCount()</td>
<td align="left">Number of aromatic bonds</td>
</tr>
<tr class="odd">
<td align="left">extractDrugAtomCount()</td>
<td align="left">Number of atom descriptor</td>
</tr>
<tr class="even">
<td align="left">extractDrugAutocorrelationCharge()</td>
<td align="left">Moreau-Broto autocorrelation descriptors using partial
charges</td>
</tr>
<tr class="odd">
<td align="left">extractDrugAutocorrelationMass()</td>
<td align="left">Moreau-Broto autocorrelation descriptors using atomic
weight</td>
</tr>
<tr class="even">
<td align="left">extractDrugAutocorrelationPolarizability()</td>
<td align="left">Moreau-Broto autocorrelation descriptors using
polarizability</td>
</tr>
<tr class="odd">
<td align="left">extractDrugBCUT()</td>
<td align="left">BCUT, the eigenvalue based descriptor</td>
</tr>
<tr class="even">
<td align="left">extractDrugBondCount()</td>
<td align="left">Number of bonds of a certain bond order</td>
</tr>
<tr class="odd">
<td align="left">extractDrugBPol()</td>
<td align="left">Sum of the absolute value of the difference between
atomic polarizabilities of all bonded atoms in the molecule</td>
</tr>
<tr class="even">
<td align="left">extractDrugCarbonTypes()</td>
<td align="left">Topological descriptor characterizing the carbon
connectivity in terms of hybridization</td>
</tr>
<tr class="odd">
<td align="left">extractDrugChiChain()</td>
<td align="left">Kier &amp; Hall Chi chain indices of orders 3, 4, 5, 6
and 7</td>
</tr>
<tr class="even">
<td align="left">extractDrugChiCluster()</td>
<td align="left">Kier &amp; Hall Chi cluster indices of orders 3, 4, 5
and 6</td>
</tr>
<tr class="odd">
<td align="left">extractDrugChiPath()</td>
<td align="left">Kier &amp; Hall Chi path indices of orders 0 to 7</td>
</tr>
<tr class="even">
<td align="left">extractDrugChiPathCluster()</td>
<td align="left">Kier &amp; Hall Chi path cluster indices of orders 4, 5
and 6</td>
</tr>
<tr class="odd">
<td align="left">extractDrugCPSA()</td>
<td align="left">Descriptors combining surface area and partial charge
information</td>
</tr>
<tr class="even">
<td align="left">extractDrugDescOB()</td>
<td align="left">Molecular descriptors provided by OpenBabel</td>
</tr>
<tr class="odd">
<td align="left">extractDrugECI()</td>
<td align="left">Eccentric connectivity index descriptor</td>
</tr>
<tr class="even">
<td align="left">extractDrugFMF()</td>
<td align="left">FMF descriptor</td>
</tr>
<tr class="odd">
<td align="left">extractDrugFragmentComplexity()</td>
<td align="left">Complexity of a system</td>
</tr>
<tr class="even">
<td align="left">extractDrugGravitationalIndex()</td>
<td align="left">Mass distribution of the molecule</td>
</tr>
<tr class="odd">
<td align="left">extractDrugHBondAcceptorCount()</td>
<td align="left">Number of hydrogen bond acceptors</td>
</tr>
<tr class="even">
<td align="left">extractDrugHBondDonorCount()</td>
<td align="left">Number of hydrogen bond donors</td>
</tr>
<tr class="odd">
<td align="left">extractDrugHybridizationRatio()</td>
<td align="left">Molecular complexity in terms of carbon hybridization
states</td>
</tr>
<tr class="even">
<td align="left">extractDrugIPMolecularLearning()</td>
<td align="left">Ionization potential</td>
</tr>
<tr class="odd">
<td align="left">extractDrugKappaShapeIndices()</td>
<td align="left">Kier &amp; Hall Kappa molecular shape indices</td>
</tr>
<tr class="even">
<td align="left">extractDrugKierHallSmarts()</td>
<td align="left">Number of occurrences of the E-State fragments</td>
</tr>
<tr class="odd">
<td align="left">extractDrugLargestChain()</td>
<td align="left">Number of atoms in the largest chain</td>
</tr>
<tr class="even">
<td align="left">extractDrugLargestPiSystem()</td>
<td align="left">Number of atoms in the largest Pi chain</td>
</tr>
<tr class="odd">
<td align="left">extractDrugLengthOverBreadth()</td>
<td align="left">Ratio of length to breadth descriptor</td>
</tr>
<tr class="even">
<td align="left">extractDrugLongestAliphaticChain()</td>
<td align="left">Number of atoms in the longest aliphatic chain</td>
</tr>
<tr class="odd">
<td align="left">extractDrugMannholdLogP()</td>
<td align="left">LogP based on the number of carbons and hetero
atoms</td>
</tr>
<tr class="even">
<td align="left">extractDrugMDE()</td>
<td align="left">Molecular Distance Edge (MDE) descriptors for C, N and
O</td>
</tr>
<tr class="odd">
<td align="left">extractDrugMomentOfInertia()</td>
<td align="left">Principal moments of inertia and ratios of the
principal moments</td>
</tr>
<tr class="even">
<td align="left">extractDrugPetitjeanNumber()</td>
<td align="left">Petitjean number of a molecule</td>
</tr>
<tr class="odd">
<td align="left">extractDrugPetitjeanShapeIndex()</td>
<td align="left">Petitjean shape indices</td>
</tr>
<tr class="even">
<td align="left">extractDrugRotatableBondsCount()</td>
<td align="left">Number of non-rotatable bonds on a molecule</td>
</tr>
<tr class="odd">
<td align="left">extractDrugRuleOfFive()</td>
<td align="left">Number failures of the Lipinski’s Rule Of Five</td>
</tr>
<tr class="even">
<td align="left">extractDrugTPSA()</td>
<td align="left">Topological Polar Surface Area (TPSA)</td>
</tr>
<tr class="odd">
<td align="left">extractDrugVABC()</td>
<td align="left">Volume of a molecule</td>
</tr>
<tr class="even">
<td align="left">extractDrugVAdjMa()</td>
<td align="left">Vertex adjacency information of a molecule</td>
</tr>
<tr class="odd">
<td align="left">extractDrugWeight()</td>
<td align="left">Total weight of atoms</td>
</tr>
<tr class="even">
<td align="left">extractDrugWeightedPath()</td>
<td align="left">Weighted path (Molecular ID)</td>
</tr>
<tr class="odd">
<td align="left">extractDrugWHIM()</td>
<td align="left">Holistic descriptors described by Todeschini et
al.</td>
</tr>
<tr class="even">
<td align="left">extractDrugWienerNumbers()</td>
<td align="left">Wiener path number and wiener polarity number</td>
</tr>
<tr class="odd">
<td align="left">extractDrugXLogP()</td>
<td align="left">Prediction of logP based on the atom-type method called
XLogP</td>
</tr>
<tr class="even">
<td align="left">extractDrugZagrebIndex()</td>
<td align="left">Sum of the squared atom degrees of all heavy atoms</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="molecular-fingerprints">Molecular fingerprints<a class="anchor" aria-label="anchor" href="#molecular-fingerprints"></a>
</h2>
<table class="table">
<caption>Table 8: Molecular fingerprints</caption>
<thead><tr class="header">
<th align="left">Function name</th>
<th align="left">Fingerprint type</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">extractDrugStandard()</td>
<td align="left">Standard molecular fingerprints (in compact
format)</td>
</tr>
<tr class="even">
<td align="left">extractDrugStandardComplete()</td>
<td align="left">Standard molecular fingerprints (in complete
format)</td>
</tr>
<tr class="odd">
<td align="left">extractDrugExtended()</td>
<td align="left">Extended molecular fingerprints (in compact
format)</td>
</tr>
<tr class="even">
<td align="left">extractDrugExtendedComplete()</td>
<td align="left">Extended molecular fingerprints (in complete
format)</td>
</tr>
<tr class="odd">
<td align="left">extractDrugGraph()</td>
<td align="left">Graph molecular fingerprints (in compact format)</td>
</tr>
<tr class="even">
<td align="left">extractDrugGraphComplete()</td>
<td align="left">Graph molecular fingerprints (in complete format)</td>
</tr>
<tr class="odd">
<td align="left">extractDrugHybridization()</td>
<td align="left">Hybridization molecular fingerprints (in compact
format)</td>
</tr>
<tr class="even">
<td align="left">extractDrugHybridizationComplete()</td>
<td align="left">Hybridization molecular fingerprints (in complete
format)</td>
</tr>
<tr class="odd">
<td align="left">extractDrugMACCS()</td>
<td align="left">MACCS molecular fingerprints (in compact format)</td>
</tr>
<tr class="even">
<td align="left">extractDrugMACCSComplete()</td>
<td align="left">MACCS molecular fingerprints (in complete format)</td>
</tr>
<tr class="odd">
<td align="left">extractDrugEstate()</td>
<td align="left">E-State molecular fingerprints (in compact format)</td>
</tr>
<tr class="even">
<td align="left">extractDrugEstateComplete()</td>
<td align="left">E-State molecular fingerprints (in complete
format)</td>
</tr>
<tr class="odd">
<td align="left">extractDrugPubChem()</td>
<td align="left">PubChem molecular fingerprints (in compact format)</td>
</tr>
<tr class="even">
<td align="left">extractDrugPubChemComplete()</td>
<td align="left">PubChem molecular fingerprints (in complete
format)</td>
</tr>
<tr class="odd">
<td align="left">extractDrugKR()</td>
<td align="left">KR (Klekota and Roth) molecular fingerprints (in
compact format)</td>
</tr>
<tr class="even">
<td align="left">extractDrugKRComplete()</td>
<td align="left">KR (Klekota and Roth) molecular fingerprints (in
complete format)</td>
</tr>
<tr class="odd">
<td align="left">extractDrugShortestPath()</td>
<td align="left">Shortest Path molecular fingerprints (in compact
format)</td>
</tr>
<tr class="even">
<td align="left">extractDrugShortestPathComplete()</td>
<td align="left">Shortest Path molecular fingerprints (in complete
format)</td>
</tr>
<tr class="odd">
<td align="left">extractDrugOBFP2()</td>
<td align="left">FP2 molecular fingerprints</td>
</tr>
<tr class="even">
<td align="left">extractDrugOBFP3()</td>
<td align="left">FP3 molecular fingerprints</td>
</tr>
<tr class="odd">
<td align="left">extractDrugOBFP4()</td>
<td align="left">FP4 molecular fingerprints</td>
</tr>
<tr class="even">
<td align="left">extractDrugOBMACCS()</td>
<td align="left">MACCS molecular fingerprints</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="protein-protein-and-compound-protein-interation-descriptors">Protein-protein and compound-protein interation descriptors<a class="anchor" aria-label="anchor" href="#protein-protein-and-compound-protein-interation-descriptors"></a>
</h2>
<table class="table">
<caption>Table 9: Protein-protein and compound-protein interation
descriptors</caption>
<thead><tr class="header">
<th align="left">Function name</th>
<th align="left">Function description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">getPPI()</td>
<td align="left">Generating protein-protein interaction descriptors</td>
</tr>
<tr class="even">
<td align="left">getCPI()</td>
<td align="left">Generating compound-protein interaction
descriptors</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="similarity-and-similarity-searching">Similarity and similarity searching<a class="anchor" aria-label="anchor" href="#similarity-and-similarity-searching"></a>
</h2>
<table class="table">
<caption>Table 10: Similarity and similarity searching</caption>
<thead><tr class="header">
<th align="left">Function name</th>
<th align="left">Function description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">calcDrugFPSim()</td>
<td align="left">Calculate drug molecule similarity derived by molecular
fingerprints</td>
</tr>
<tr class="even">
<td align="left">calcDrugMCSSim()</td>
<td align="left">Calculate drug molecule similarity derived by maximum
common substructure search</td>
</tr>
<tr class="odd">
<td align="left">searchDrug()</td>
<td align="left">Parallelized drug molecule similarity search by
molecular fingerprints similarity or maximum common substructure
search</td>
</tr>
<tr class="even">
<td align="left">calcTwoProtSeqSim()</td>
<td align="left">Similarity calculation based on sequence alignment for
a pair of protein sequences</td>
</tr>
<tr class="odd">
<td align="left">calcParProtSeqSim()</td>
<td align="left">Parallellized protein sequence similarity calculation
based on sequence alignment</td>
</tr>
<tr class="even">
<td align="left">calcTwoProtGOSim()</td>
<td align="left">Similarity calculation based on Gene Ontology (GO)
similarity between two proteins</td>
</tr>
<tr class="odd">
<td align="left">calcParProtGOSim()</td>
<td align="left">Protein similarity calculation based on Gene Ontology
(GO) similarity</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="protein-sequence-data-manipulation">Protein sequence data manipulation<a class="anchor" aria-label="anchor" href="#protein-sequence-data-manipulation"></a>
</h2>
<table class="table">
<caption>Table 11: Protein sequence data manipulation</caption>
<thead><tr class="header">
<th align="left">Function name</th>
<th align="left">Function description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">readFASTA()</td>
<td align="left">Read protein sequences in FASTA format</td>
</tr>
<tr class="even">
<td align="left">readPDB()</td>
<td align="left">Read protein sequences in PDB format</td>
</tr>
<tr class="odd">
<td align="left">segProt()</td>
<td align="left">Protein sequence segmentation</td>
</tr>
<tr class="even">
<td align="left">checkProt()</td>
<td align="left">Check if the protein sequence’s amino acid types are
the 20 default types</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="molecular-data-manipulation">Molecular data manipulation<a class="anchor" aria-label="anchor" href="#molecular-data-manipulation"></a>
</h2>
<table class="table">
<caption>Table 12: Molecular data manipulation</caption>
<thead><tr class="header">
<th align="left">Function name</th>
<th align="left">Function description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">readMolFromSDF()</td>
<td align="left">Read molecules from SDF files and return parsed Java
molecular object</td>
</tr>
<tr class="even">
<td align="left">readMolFromSmi()</td>
<td align="left">Read molecules from SMILES files and return parsed Java
molecular object or plain text list</td>
</tr>
<tr class="odd">
<td align="left">convMolFormat()</td>
<td align="left">Chemical file formats conversion</td>
</tr>
</tbody>
</table>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Nan Xiao, Dong-Sheng Cao, Qing-Song Xu.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
